kernel/kernel-api/src/main/java/io/delta/kernel/internal/replay/LogReplay.java

```diff
@@ -392,4 +372,52 @@ private Map<String, DomainMetadata> loadDomainMetadataMap(Engine engine) {
       throw new UncheckedIOException("Could not close iterator", ex);
     }
   }
+
+  /**
+   * Calculates the latest snapshot hint before or at the current snapshot version, returns the
+   * CRCInfo if checksum file at the current version is read
+   */
+  private Tuple2<Optional<SnapshotHint>, Optional<CRCInfo>>
+      maybeGetNewerSnapshotHintAndCurrentCrcInfo(
+          Engine engine,
+          LogSegment logSegment,
+          Optional<SnapshotHint> snapshotHint,
+          long snapshotVersion) {
+
+    // Snapshot hint's version is current.
+    if (snapshotHint.isPresent() && snapshotHint.get().getVersion() == snapshotVersion) {
+      return new Tuple2<>(snapshotHint, Optional.empty());
+    }
+
+    // Ignore the snapshot hint whose version is larger.
+    if (snapshotHint.isPresent() && snapshotHint.get().getVersion() > snapshotVersion) {
+      snapshotHint = Optional.empty();
+    }
+
+    long crcSearchLowerBound =
+        max(
+            asList(
+                // Prefer reading hint over CRC, so start listing from hint's version + 1,
+                // if hint is not present, list from version 0.
+                snapshotHint.map(SnapshotHint::getVersion).orElse(-1L) + 1,
```

Comment:
After making change below, I observe a corner case of hit missing, read 0 will fail the check
checkArgument(crcInfo.getVersion() >= crcSearchLowerBound && crcInfo.getVersion() <= snapshotVersion);

Fix it.

(Old implementation is also correct because this check won't happen when version = 0 and find 0.crc)